dcos_tunnel/cli.py

```diff
@@ -632,37 +654,49 @@ def _vpn(port, config_file, user, privileged, ssh_port, host,
 
         msg = "\nWaiting for VPN server in container '{}' to come up..."
         emitter.publish(msg.format(container_name))
+
         scom = ("until "
-                """[ "$(docker inspect --format='{{ .State.Running }}' """
+                """[ "$(%s inspect --format='{{ .State.Running }}' """
                 """%s 2>/dev/null)" = "true" ] 2>/dev/null; do sleep 0.5; """
-                """done""") % (container_name)
-        scom += (";"
-                 """docker exec {} "sh -c 'until [ -s {} ]; do sleep 0.5; """
-                 """done'" """).format(container_name, remote_keyfile)
-        scom += (";"
-                 """docker exec {} "sh -c 'until [ -s {} ]; do sleep 0.5; """
-                 """done'" """).format(container_name, remote_clientfile)
+                """done""") % (docker_cmd, container_name)
+        scom += (" && "
+                 """{} exec {} sh -c 'until [ -s {} ]; do sleep 0.5; """
+                 """done' """).format(docker_cmd, container_name,
+                                      remote_keyfile)
+        scom += (" && "
+                 """{} exec {} sh -c 'until [ -s {} ]; do sleep 0.5; """
+                 """done' """).format(docker_cmd, container_name,
+                                      remote_clientfile)
         server_setup = client.exec_command(scom, get_pty=True)
-        ssh_exec_wait(server_setup)
 
-        scom = ("""docker inspect --format='"""
+        if not vpn_server.is_alive():
```

Comment:
The hanging behavior you observed, where the vpn server never came up and the wait message was hanging forever, was because of the `until` loop inside the `scom`, not because the thread terminated. The case where the container never comes up because the docker command was invalid is caught by the new checking logic so it won't happen again.

The only way to prevent the hanging here is to introduce a timeout. But I'm not sure if that would be helpful here, as the user can just terminate the command on their own.